models/weight_recorder.cpp

```diff
@@ -0,0 +1,170 @@
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  NEST is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with NEST.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "weight_recorder.h"
+
+// C++ includes:
+#include <numeric>
+
+// Includes from libnestutil:
+#include "compose.hpp"
+#include "logging.h"
+
+// Includes from nestkernel:
+#include "event_delivery_manager_impl.h"
+#include "kernel_manager.h"
+#include "sibling_container.h"
+
+// Includes from sli:
+#include "arraydatum.h"
+#include "dict.h"
+#include "dictutils.h"
+#include "doubledatum.h"
+#include "integerdatum.h"
+
+nest::weight_recorder::weight_recorder()
+  : Node()
+  // record time, gid and weight
+  , device_( *this,
+      RecordingDevice::SPIKE_DETECTOR,
+      "gdf",
+      true,
+      true,
+      true,
+      true )
+  , user_set_precise_times_( false )
+  , has_proxies_( false )
+  , local_receiver_( true )
+{
+}
+
+nest::weight_recorder::weight_recorder( const weight_recorder& n )
+  : Node( n )
+  , device_( *this, n.device_ )
+  , user_set_precise_times_( n.user_set_precise_times_ )
+  , has_proxies_( false )
+  , local_receiver_( true )
+{
+}
+
+void
+nest::weight_recorder::init_state_( const Node& np )
+{
+  const weight_recorder& wr = dynamic_cast< const weight_recorder& >( np );
+  device_.init_state( wr.device_ );
+  init_buffers_();
+}
+
+void
+nest::weight_recorder::init_buffers_()
+{
+  device_.init_buffers();
+  B_.events_ = std::vector< WeightRecorderEvent >();
+}
+
+void
+nest::weight_recorder::calibrate()
+{
+  if ( kernel().event_delivery_manager.get_off_grid_communication()
+    and not device_.is_precise_times_user_set() )
+  {
+    device_.set_precise_times( true );
+    std::string msg = String::compose(
+      "Precise neuron models exist: the property precise_times "
+      "of the %1 with gid %2 has been set to true",
+      get_name(),
+      get_gid() );
+
+    if ( device_.is_precision_user_set() )
+    {
+      // if user explicitly set the precision, there is no need to do anything.
+      msg += ".";
+    }
+
+    else
+    {
+      // it makes sense to increase the precision if precise models are used.
+      device_.set_precision( 15 );
+      msg += ", precision has been set to 15.";
+    }
+
+    LOG( M_INFO, "spike_detector::calibrate", msg );
+  }
+
+  device_.calibrate();
+}
+
+void
+nest::weight_recorder::update( Time const&, const long from, const long to )
+{
+
+  for ( std::vector< WeightRecorderEvent >::iterator e = B_.events_.begin();
+        e != B_.events_.end();
+        ++e )
+  {
```

Comment:
So, for example, couldn't we test here whether the sender is one of the desired presynaptic population (if specified), ditto receiver, and only record it if it was?